README
[phZoom] 是一个小巧的jQuery看图插件.
==================================
当前版本: phZoom 1.292 Final
更新时间: 2012.1.6

使用方法:
1. 加载jQuery库(需1.4+)
2. 加载phZoom.css, 请保证压缩包内的两张小图片与phzoom.css同级目录
3. 加载phzoom.js, 请保证其加载于jQuery库之后
4. 调用接口:
$node.phzoom({
// 此处为设置项, 可留空
});
请把 $node 改为所需的jQuery选择器, 如: $('a:has(img)')

不管如何调用, 请务必将调用代码包在一个DOM Ready方法内, 如:
$(function(){
// 此处为上面的调用接口
});
以下为插件设置选项, 大括号结束前不要逗号
{
	layOpacity:0.7,          // 遮罩层不透明度应为0到1之间的某个小数, 默认0.7即为70%不透明
	layDur:300,              // 遮罩层渐显的时间, 纯数字, 单位毫秒
	animDurA:300,            // 图片从原位飞到荧幕中间的时间, 纯数字, 单位毫秒
	animDurB:300,      		 // 图片从小到大的变化时间, 纯数字, 单位毫秒
	navColor:'#cf0',         // 上/下一张按钮的文本颜色, 需要引号
	capColor:'#cf0',         // 大图底部标题与索引的文本颜色, 需要引号
	prevText:'Prev',         // 上一张按钮的显示文本, 需要引号
	nextText:'Next',         // 下一张按钮的显示文本, 需要引号
	limitWidth:false,        // 默认为false, 大图使用原始宽度; 若设为true, 特大图将不会宽于窗口
	returnOrigin:true        // 默认为true, 此时返回初始对象集合; 若设为false, 则仅返回包含img元素的对象集合
}


更新历史:
2012.1.6 | 1.292 Final
1. 继续强化pos数组, 提高其效率
2. 调整了一些重要变量的命名

2012.1.5 | 1.291 Final
1. 修复IE(6|7|8)下无法创建小图hover的BUG
2. 设置项returnOrigin现在可以完全正常工作了

2012.1.2 | 1.29 Final
1. 精简pos数组, 减少两个item
2. 略微提升大图loading阶段的性能
3. 希望1.30将是phZoom的第一个稳定版

2011.12.22 | 1.28 Final
1. 现在完全无需使用 :has(img) 选择器来调用phZoom了
2. 新增设置项: prevText和nextText
3. 新增设置项: returnOrigin
4. 其它一些无关性能的细节改进

2011.12.17 | 1.27 Final
1. 修复当原图极大时可能溢出容器100%宽度的BUG

2011.12.11 | 1.26 Final
1. 修复IE(6|7|8)下导致新窗口打开的BUG

2011.12.06 | 1.25 Final
1. 使用jQuery.data缓存对象
2. 一些命名调整与细节优化
3. 终于正式版了

2011.11.05 | 1.23 RC
1. 修正某些页面大图无法居中的问题
2. 中止载入后, 大图不会再自动弹出了
3. 防止重复生成遮罩层与大图层
4. 其它一些细节完善

2011.10.18 | Beta 1.22
1. 修复新版Chrome下无法正确读取缓存图片尺寸的BUG
2. 精简掉几个变量与方法, 改用接力棒法传递参数

2011.10.15 | Beta 1.21
1. 修复IE6下Body高度过小时的遮罩层问题

2011.09.28 | Beta 1.20
0. 插件基本重写, 并追加了代码注释
1. 加快图片载入速度, 提高插件执行效率
2. 点击图片不再是单一的退出大图(自己体验)
3. 上/下一张的操作变得相当自由了
4. 遮罩层改用固定定位
5. Firefox下不会卡了
6. 对老版Opera作了兼容
7. 去除了防迷路模式

2011.07.15 | Beta 1.02
1. 查看特大图超出荧幕部分时使用最小量动画
2. 修复一个Opera下的神奇BUG

2011.07.14
1. 修复大图Caption为空时的BUG
2. 现在大图展示更流畅了
3. 更精确的大图层高度

2011.07.13 | Beta 1.01
1. 修复IE(6|7|8)下一个函数执行顺序的BUG
2. 优化大图标题创建方式
3. 预加载前一张与后一张


如果你对这个插件有兴趣, 
敬请关注作者的博客:
http://phoetry.me
插件发布页面:
http://phoetry.me/archives/phzoom.html
GitHub页面:
https://github.com/phoetry/phZoom

感谢你的支持与反馈.
作者: Phoetry

changelog.md
## phZoom 更新历史
* * *
2012.1.6 | 1.292 Final  
1. 继续强化pos数组, 提高其效率  
2. 使用YUICompressor压缩代码  
3. 调整了一些重要变量的命名

2012.1.5 | 1.291 Final  
1. 修复IE(6|7|8)下无法创建小图hover的BUG  
2. 设置项 `returnOrigin` 现在可以完全正常工作了

2012.1.2 | 1.29 Final  
1. 精简pos数组, 减少两个item  
2. 略微提升大图loading阶段的性能  
3. 希望1.30将是phZoom的第一个稳定版

2011.12.22 | 1.28 Final  
1. 现在完全无需使用 :has(img) 选择器来调用phZoom了  
2. 新增设置项: `prevText` 和 `nextText`  
3. 新增设置项: `returnOrigin`  
4. 其它一些无关性能的细节改进

2011.12.17 | 1.27 Final  
1. 修复当原图极大时可能溢出容器100%宽度的BUG

2011.12.11 | 1.26 Final  
1. 修复IE(6|7|8)下导致新窗口打开的BUG

2011.12.06 | 1.25 Final  
1. 使用 `jQuery.data` 缓存对象  
2. 一些命名调整与细节优化  
3. 终于正式版了

2011.11.05 | 1.23 RC  
1. 修正某些页面大图无法居中的问题  
2. 中止载入后, 大图不会再自动弹出了  
3. 防止重复生成遮罩层与大图层  
4. 其它一些细节完善

2011.10.18 | Beta 1.22  
1. 修复新版Chrome下无法正确读取缓存图片尺寸的BUG  
2. 精简掉几个变量与方法, 改用接力棒法传递参数

2011.10.15 | Beta 1.21  
1. 修复IE6下Body高度过小时的遮罩层问题

2011.09.28 | Beta 1.20  
0. 插件基本重写, 并追加了代码注释  
1. 加快图片载入速度, 提高插件执行效率  
2. 点击图片不再是单一的退出大图(自己体验)  
3. 上/下一张的操作变得相当自由了  
4. 遮罩层改用固定定位  
5. Firefox下不会卡了  
6. 对老版Opera作了兼容  
7. 去除了防迷路模式

2011.07.15 | Beta 1.02  
1. 查看特大图超出荧幕部分时使用最小量动画  
2. 修复一个Opera下的神奇BUG

2011.07.14 | Beta 1.011
1. 修复大图Caption为空时的BUG  
2. 现在大图展示更流畅了  
3. 更精确的大图层高度

2011.07.13 | Beta 1.01  
1. 修复IE(6|7|8)下一个函数执行顺序的BUG  
2. 优化大图标题创建方式  
3. 预加载前一张与后一张

2011.07.10 | Beta 1.0  
1. 最初发布测试版

phzoom.js
~function($){var W=$(window),D=$(document),LAY=$('<div id=ph_lay/>'),ZOOM=$('<div id=ph_zoom/>'),BOTH=LAY.add(ZOOM),PHZOOM=function(e,x,y,z){this.opt=x;this.idx=y;this.all=z;this.len=z.length;this.end=this.len>y+1;this.img=$('img:first',e);this.lnk=e.addClass('phzoom').unbind('click').bind(this.imgFn()).append(this.hov=$('<span class=ph_hover/>').hide())[0];this.cap=$('<div/>',{css:{color:x.capColor},id:'ph_cap',html:$([$('<span/>',{id:'ph_txt',text:this.img[0].title||this.lnk.title||'No title'})[0],$('<span/>',{id:'ph_idx',text:y+1+' / '+this.len})[0]])}).add(this.nav=$('<div/>',{id:'ph_nav',css:{color:x.navColor},html:(y?'<span id=ph_prev>'+x.prevText+'</span>':'')+(this.end?'<span id=ph_next>'+x.nextText+'</span>':'')}));BOTH.click($.proxy(this,'imgQuit'));window.XMLHttpRequest||e.height(this.img.height())};PHZOOM.prototype={imgFn:function(){var that=this,s=function(){return that.hov.not('.loading').stop(0,1)};return{mouseover:function(){s().fadeIn()},mouseout:function(){s().fadeOut()},click:function(){that.imgLoad();return false}}},imgPos:function(oW,oH){var A=this.img,L=W.scrollLeft(),T=W.scrollTop(),pos=[W.width(),W.height(),A.width(),A.height(),A.offset().left,A.offset().top];this.opt.limitWidth&&oW>pos[0]&&(oH=oH/oW*(oW=pos[0]));return pos.concat(oW,oH,(pos[0]-oW)/2+L,(pos[1]-oH)/2+T,(pos[0]-pos[2])/2+L,(pos[1]-pos[3])/2+T)},imgLoad:function(){LAY.fadeTo(this.opt.layDur,this.opt.layOpacity);var that=this,B=new Image;this.hov.addClass('loading');B.className='zoomed';B.onload=function(){B.onload=null;that.hov.hasClass('loading')&&(ZOOM.height(D.height()).append(B).show(),that.imgAnim(B),that.preLoad())};B.src=this.lnk.href},imgAnim:function(B){var that=this,$B=$(B),pos=this.imgPos(B.width||+$B.attr('width'),B.height||+$B.attr('height')),oFlow=pos[0]<pos[6],E=this.evtMon($B,pos[0],pos[0]-pos[6],!oFlow);$B.after(this.cap.hide()).css({left:pos[4],top:pos[5],width:pos[2],height:pos[3]}).animate({left:pos[10],top:pos[11]},this.opt.animDurA,function(){$B.animate({left:pos[8],top:pos[9],width:pos[6],height:pos[7]},that.opt.animDurB,function(){that.hov.removeClass('loading');that.cap.css({top:pos[7]+pos[9],left:oFlow?0:pos[8],width:oFlow?pos[0]:pos[6]}).fadeTo(300,.7);that.nav.bind(E).css('top',pos[7]/3+pos[9]);that.keyBind()}).bind(E)})},imgQuit:function(isQuit){this.hov.hide().hasClass('loading')?this.hov.removeClass('loading'):D.unbind('.phzoom');isQuit&&LAY.fadeOut();ZOOM.hide().empty();return false},imgChange:function(num){this.imgQuit();$('.ph_hover',$(this.all[this.idx+num]).click()).show();return false},preLoad:function(x,y){this.idx&&(x=new Image,x.src=this.all[this.idx-1].href,delete x);this.end&&(y=new Image,y.src=this.all[this.idx+1].href,delete y)},keyBind:function(){var that=this;D.bind('keydown.phzoom',function(e){e=e.which;return e==37&&that.idx?that.imgChange(-1):e==39&&that.end?that.imgChange(1):e^27||that.imgQuit(1)})},evtMon:function($B,a,b,c){var that=this,nav=$('span',this.nav).hide();return{click:function(e){return that.len<2||(e=e.pageX>a/2,that.idx?that.end?that.imgChange(e||-1):e||that.imgChange(-1):!e||that.imgChange(1))},mouseout:function(){nav.hide()},mousemove:function(e,i){e=e.pageX,i=e>a/2;that.idx?(nav.eq(i).show(),nav.eq(1-i).hide()):nav[i?'show':'hide']();c||(e=e<a/3?0:e>a*2/3?b:b/2)==$B.position().left||$B.not(':animated').animate({left:e},200)}}}};$.phzoom=function(Z,x,z){x=$.extend({layOpacity:.7,layDur:300,animDurA:300,animDurB:300,navColor:'#cf0',capColor:'#cf0',prevText:'Prev',nextText:'Next',limitWidth:false,returnOrigin:true},x),(z=Z.has('img'))[0]&&($('#ph_lay')[0]||$('body').append(BOTH),z.each(function(y,t){$.data(t,'phzoom',new PHZOOM($(t),x,y,z))}));return x.returnOrigin?Z:z};$.fn.phzoom=function(x){return $.phzoom(this,x)}}(jQuery);

phzoom.src/phzoom.src.js
/**
 * @name jQuery phZoom Plugin
 * @version 1.292 Final
 * @create 2011.7.10
 * @lastmodified 2012.1.6
 * @description Based on jQuery 1.4+
 * @author Phoetry (http://phoetry.me)
 * @url http://phoetry.me/archives/phzoom.html
 **/
~function($){var
/** 
 * @param LAY:遮罩层, ZOOM:大图容器, PHZOOM:构造主函数
 * @param e:当前对象, x:插件设置, y:当前index, z:对象集合
 **/
W=$(window),D=$(document),
LAY=$('<div id=ph_lay/>'),
ZOOM=$('<div id=ph_zoom/>'),
BOTH=LAY.add(ZOOM),
PHZOOM=function(e,x,y,z){
	this.opt=x;
	this.idx=y;
	this.all=z;
	this.len=z.length;
	// 当前e为对象集合末项时:false
	this.end=this.len>y+1;
	// 获取当前e的小图(返回jQuery)
	this.img=$('img:first',e);
	// 初始化当前e并绑定事件(返回DOM)
	this.lnk=e.addClass('phzoom').unbind('click').bind(this.imgFn())
		.append(this.hov=$('<span class=ph_hover/>').hide())[0];
	// cap:大图底部的标题+索引+上/下一张(合体)
	this.cap=$('<div/>',{
		css:{color:x.capColor},
		id:'ph_cap',
		html:$([
			$('<span/>',{//标题
				id:'ph_txt',
				text:this.img[0].title||this.lnk.title||'No title'
			})[0],
			$('<span/>',{//索引
				id:'ph_idx',
				text:y+1+' / '+this.len
			})[0]
		])
	}).add(
	// nav:上/下一张
	this.nav=$('<div/>',{
		id:'ph_nav',
		css:{color:x.navColor},
		html:(y?'<span id=ph_prev>'+x.prevText+'</span>':'')
			+(this.end?'<span id=ph_next>'+x.nextText+'</span>':'')
	}));
	// 点击页面上随便哪里都能退出(后面会排除大图区域)
	BOTH.click($.proxy(this,'imgQuit'));
	// 尝试Fix IE6下hover可能错位的BUG
	window.XMLHttpRequest||e.height(this.img.height());
};
/** 
 * prototype of PHZOOM function
 * @param B:new Image, $B:B的jQuery对象
 **/
PHZOOM.prototype={
	// 当前对象绑定的事件
	imgFn:function(){
		var that=this,
			s=function(){
				return that.hov.not('.loading').stop(0,1);
			};
		return{
			mouseover:function(){s().fadeIn()},
			mouseout:function(){s().fadeOut()},
			click:function(){
				that.imgLoad();
				return false;
			}
		}
	},
	// 为之后动画准备一些必需品, A:小图(jQuery), oW:大图原始宽度, oH:大图原始高度
	imgPos:function(oW,oH){
		var A=this.img,L=W.scrollLeft(),T=W.scrollTop(),
			pos=[
				W.width(),W.height(),
				A.width(),A.height(),
				A.offset().left,A.offset().top
			];
		// 限宽模式下自动调整特大图
		this.opt.limitWidth&&oW>pos[0]&&(oH=oH/oW*(oW=pos[0]));
		return pos.concat(
			oW,oH,
			(pos[0]-oW)/2+L,
			(pos[1]-oH)/2+T,
			(pos[0]-pos[2])/2+L,
			(pos[1]-pos[3])/2+T
		);
	},
	// 开始加载大图
	imgLoad:function(){
		LAY.fadeTo(this.opt.layDur,this.opt.layOpacity);
		var that=this,B=new Image;
		this.hov.addClass('loading');
		B.className='zoomed';
		B.onload=function(){
			B.onload=null;
			// 如果此时已经退出大图, 则停止执行
			that.hov.hasClass('loading')&&(
				// resize之类的事件会影响文档尺寸, 故height一下
				ZOOM.height(D.height()).append(B).show(),
				that.imgAnim(B),that.preLoad()
			);
		};
		B.src=this.lnk.href;
	},
	// 大图执行两次动画, 动画完毕后绑定事件
	imgAnim:function(B){
		var that=this,$B=$(B),
			// 获取动画所需的尺寸与位置, 传参:大图的原始宽度和高度
			pos=this.imgPos(
				B.width||+$B.attr('width'),
				B.height||+$B.attr('height')
			),
			// 当前大图溢出Body宽度时:true
			oFlow=pos[0]<pos[6],
			// 预备好动画后需要绑定的事件
			E=this.evtMon($B,pos[0],pos[0]-pos[6],!oFlow);
		$B.after(this.cap.hide()).css({//定位1
			left:pos[4],top:pos[5],
			width:pos[2],height:pos[3]
		}).animate({//定位2, 动画1
			left:pos[10],top:pos[11]
		},this.opt.animDurA,function(){
			$B.animate({//定位3, 动画2
				left:pos[8],top:pos[9],
				width:pos[6],height:pos[7]
			},that.opt.animDurB,function(){//动画完
				that.hov.removeClass('loading');
				that.cap.css({//定位cap
					top:pos[7]+pos[9],
					left:oFlow?0:pos[8],
					width:oFlow?pos[0]:pos[6]
				}).fadeTo(300,.7);
				//定位nav并绑定鼠标事件
				that.nav.bind(E).css('top',pos[7]/3+pos[9]);
				//绑定快捷键
				that.keyBind();
			}).bind(E);
		});
	},
	// 退出大图, isQuit为undefined时化身为imgChange的过程(将保持遮罩层)
	imgQuit:function(isQuit){
		this.hov.hide().hasClass('loading')?this.hov.removeClass('loading'):D.unbind('.phzoom');
		isQuit&&LAY.fadeOut();
		ZOOM.hide().empty();
		return false;
	},
	// 切换上/下一张, 不传参给imgQuit
	imgChange:function(num){
		this.imgQuit();
		$('.ph_hover',$(this.all[this.idx+num]).click()).show();
		return false;
	},
	// 预加载相邻图片
	preLoad:function(x,y){
		this.idx&&(x=new Image,x.src=this.all[this.idx-1].href,delete x);
		this.end&&(y=new Image,y.src=this.all[this.idx+1].href,delete y);
	},
	// 绑定快捷键, 逃脱键:退出, 左箭头:上一张, 右箭头:下一张
	keyBind:function(){
		var that=this;
		D.bind('keydown.phzoom',function(e){
			e=e.which;
			return e==37&&that.idx?that.imgChange(-1):
				e==39&&that.end?that.imgChange(1):
				e^27||that.imgQuit(1)
		});
	},
	// 大图加载完毕后需要绑定的事件
	evtMon:function($B,a,b,c){
		var that=this,nav=$('span',this.nav).hide();
		return{
			click:function(e){
				return that.len<2||(
					e=e.pageX>a/2,
					that.idx?that.end
					?that.imgChange(e||-1)
					:e||that.imgChange(-1)
					:!e||that.imgChange(1)
				);
			},
			mouseout:function(){nav.hide()},
			mousemove:function(e,i){
				e=e.pageX,i=e>a/2;
				that.idx?(
					nav.eq(i).show(),
					nav.eq(1-i).hide()
				):nav[i?'show':'hide']();
				c||(e=e<a/3?0:e>a*2/3?b:b/2)==$B.position().left||
				$B.not(':animated').animate({left:e},200);
			}
		}
	}
};
/** 
 * 直接扩展jQuery底层方法
 * @param x:插件设置项, y:当前index, Z:初始传入的对象集合, z:过滤后的对象集合
 **/
$.phzoom=function(Z,x,z){
	x=$.extend({
		layOpacity:.7,
		layDur:300,
		animDurA:300,
		animDurB:300,
		navColor:'#cf0',
		capColor:'#cf0',
		prevText:'Prev',
		nextText:'Next',
		limitWidth:false,
		returnOrigin:true
	},x),
	(z=Z.has('img'))[0]&&(
		$('#ph_lay')[0]||
		$('body').append(BOTH),
		z.each(function(y,t){
			$.data(t,'phzoom',new PHZOOM($(t),x,y,z));
		})
	);
	return x.returnOrigin?Z:z;
};
// 插件调用接口, hook me, 完毕.
$.fn.phzoom=function(x){
	return $.phzoom(this,x);
};
function log(z){
	console.log({}.toString.call(z)+' | '+z);
}
}(jQuery);